### DISCUSSION TITLE
Linting a custom def* macro that takes a vector of param bindings exactly like destructuring :keys vector

### DISCUSSION THREAD

**Discussion by TuggyNE** [clj-kondo/clj-kondo#1502] [Q&A] [answered]
I have a custom macro that takes a vector of parameter bindings and injects this into the vector of a `{:keys[]}` destructuring parameter form, as well as using it in a generated `(s/keys :req [])` spec. Linting it as `defmethod` almost works (somewhat coincidentally, it has the right arrangement of macro parameters), except for one special case that is unfortunately unavoidable and shows up as errors in clj-kondo.

Specifically, it's perfectly permissible to include keywords in the vector, even namespaced, and it works exactly like you'd expect with normal destructuring, creating a local binding with the `name` of the keyword. Unfortunately, `defmethod` obviously does not support this and so it lints wrong.

On the other hand, linting as `clj-kondo.lint-as/def-catch-all` is not ideal either, since it ends up creating some spurious unused var warnings, and potentially hiding some legitimate undefined var errors too: the macro does not define _every_ symbol its form includes, only some of them.

Is there a reasonably straightforward solution for this? (I suppose I could theoretically rewrite the macro to use less syntactic sugar and require the actual destructuring form, but then I'd have to change two or three dozen places, and the macro would need more internal sophistication.)

**Comment by borkdude** [comment, marked as answer]
@TuggyNE My suggestion for this macro would be to write a hook for it. The easiest to get started with is the `:macroexpand` hook: https://github.com/clj-kondo/clj-kondo/blob/master/doc/hooks.md#macroexpand

You can simply copy past your macro, or a simplified version into a clojure file in the config dir: `.clj-kondo/your-org/your-lib/your_lib/your_namespace.clj`
And then configure that calls to your macro should use that hook.

**Reply by TuggyNE** [reply]
Thanks! I got this to work OK from the clj-kondo CLI with two different subdirectory arrangements under .clj-kondo (one with a path prefix of `/dev.company-name/project-name`, and the other with only the trailing part of `/dev/company_name/project_name/file.clj`), but I can't quite figure out how to structure the files to match the nses for Calva's integration (so the hook apparently is not found). Do you know anything about that or should I ask over there?

I checked the clojure-lsp server log and found one place where the proper file evidently wasn't found, but further attempts to trigger re-checks didn't create any more error messages of that sort.

Then I restarted VS Code and after clojure-lsp restarted, it didn't show further errors about not finding the hook, but it still didn't lint using it correctly.

**Comment by TuggyNE** [comment]
All right, finally got this all to work by using the no-prefix path (`.clj-kondo/hooks/dev/company_name/project_name/file.clj` and putting `config.edn` in `.clj-kondo` directly rather than in the prefix directory (`.clj-kondo/dev.company-name/project-name`), then restarting to make the changes visible.

**Reply by borkdude** [reply]
@TuggyNE Glad you figured it out. Starting with the newest clj-kondo version released two days ago, all configurations in 

```
.clj-kondo/*/*/config.edn
```
are automatically picked up.

**Reply by borkdude** [reply]
Also note that when referring to hook code, the namespaces are always relative to the `config.edn` file.
